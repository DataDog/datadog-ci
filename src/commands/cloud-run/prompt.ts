import inquirer from 'inquirer'

import {DATADOG_SITES} from '../../constants'

const checkboxPlusPrompt = require('inquirer-checkbox-plus-prompt')
inquirer.registerPrompt('checkbox-plus', checkboxPlusPrompt)

export const requestGCPProject = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    message: 'Enter GCP Project ID:',
    name: 'project',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Project ID is required.'
      }

      return true
    },
  })

  return answer.project
}

export const requestGCPRegion = async (defaultRegion?: string): Promise<string> => {
  const answer = await inquirer.prompt({
    default: defaultRegion || 'us-central1',
    message: 'Enter GCP Region:',
    name: 'region',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Region is required.'
      }

      return true
    },
  })

  return answer.region
}

export const requestServiceName = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    message: 'Enter Cloud Run service name:',
    name: 'serviceName',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Service name is required.'
      }

      return true
    },
  })

  return answer.serviceName
}

export const requestSite = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    choices: DATADOG_SITES,
    message: 'Select a Datadog Site:',
    name: 'site',
    type: 'list',
  })

  return answer.site
}

export const requestConfirmation = async (message: string, defaultValue = true) => {
  const confirmationAnswer = await inquirer.prompt({
    message,
    name: 'confirmation',
    type: 'confirm',
    default: defaultValue,
  })

  return confirmationAnswer.confirmation !== false
}
