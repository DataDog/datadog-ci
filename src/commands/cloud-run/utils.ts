import type {IService, ServicesClient as IServicesClient} from './types'

import chalk from 'chalk'
import {GoogleAuth} from 'google-auth-library'
import {diff} from 'jest-diff'

import {withSpinner} from './renderer'

/**
 * Check if the user is authenticated with GCP.
 * @returns true if the user is authenticated, false otherwise
 */
export const checkAuthentication = async () => {
  const auth = new GoogleAuth()
  try {
    await auth.getApplicationDefault()

    return true
  } catch (_) {
    return false
  }
}

/**
 * Recursively sort object keys to ensure consistent ordering
 */
const sortObjectKeys = (obj: any): any => {
  if (!obj) {
    return obj
  }

  if (Array.isArray(obj)) {
    return obj.map(sortObjectKeys)
  }

  if (typeof obj === 'object') {
    const sorted: any = {}
    Object.keys(obj)
      .sort()
      .forEach((key) => {
        sorted[key] = sortObjectKeys(obj[key])
      })

    return sorted
  }

  return obj
}

/**
 * Obfuscate sensitive values in a line if it contains a key with "_KEY"
 */
const obfuscateSensitiveValues = (line: string): string => {
  // Match hex strings of 16, 32, or 64 characters (common API key/token lengths)
  return line
    .replace(/("[0-9a-fA-F]{16}"|"[0-9a-fA-F]{32}"|"[0-9a-fA-F]{64}")/g, '"***"')
    .replace(/('[0-9a-fA-F]{16}'|'[0-9a-fA-F]{32}'|'[0-9a-fA-F]{64}')/g, "'***'")
}

/**
 * Generate a git diff-style comparison between two configurations
 * TODO(@nhulston): update Lambda and AAS instrument to show this diff
 * @param original The original configuration object
 * @param updated The updated configuration object
 * @returns A formatted diff string with colors
 */
export const generateConfigDiff = (original: any, updated: any): string => {
  // Sort keys consistently before comparison
  const sortedOriginal = sortObjectKeys(original)
  const sortedUpdated = sortObjectKeys(updated)

  const originalJson = JSON.stringify(sortedOriginal, undefined, 2)
  const updatedJson = JSON.stringify(sortedUpdated, undefined, 2)

  const obfuscatedOriginal = originalJson.split('\n').map(obfuscateSensitiveValues).join('\n')
  const obfuscatedUpdated = updatedJson.split('\n').map(obfuscateSensitiveValues).join('\n')

  const configDiff = diff(obfuscatedOriginal, obfuscatedUpdated, {
    aColor: chalk.red,
    bColor: chalk.green,
    omitAnnotationLines: true,
  })
  if (!configDiff || configDiff.includes('no visual difference')) {
    return chalk.gray('No changes detected.')
  }

  return configDiff
}

export const fetchServiceConfigs = async (
  client: IServicesClient,
  project: string,
  region: string,
  services: string[]
) => {
  const existingServiceConfigs: IService[] = []
  for (const serviceName of services) {
    const servicePath = client.servicePath(project, region, serviceName)

    const existingService = await withSpinner(
      `Fetching configuration for ${chalk.bold(serviceName)}...`,
      async () => {
        try {
          const [serv] = await client.getService({name: servicePath})

          return serv
        } catch (error) {
          throw new Error(
            `Service ${serviceName} not found in project ${project}, region ${region}.\n\nNo services were instrumented.\n`
          )
        }
      },
      `Fetched service configuration for ${chalk.bold(serviceName)}`
    )
    existingServiceConfigs.push(existingService)
  }

  return existingServiceConfigs
}
