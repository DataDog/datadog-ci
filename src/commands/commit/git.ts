import * as simpleGit from 'simple-git'
import {Writable} from 'stream'
import {URL} from 'url'
import {renderGitError} from './renderer'

import {CommitInfo} from './interfaces'

// Returns a configured SimpleGit.
export const newSimpleGit = async (): Promise<simpleGit.SimpleGit> => {
  const options = {
    baseDir: process.cwd(),
    binary: 'git',
    maxConcurrentProcesses: 1,
  }
  try {
    // Attempt to set the baseDir to the root of the repository so the 'git ls-files' command
    // returns the tracked files paths relative to the root of the repository.
    const git = simpleGit.gitP(options)
    const root = await git.revparse('--show-toplevel')
    options.baseDir = root
  } catch {
    // Ignore exception as it will fail if we are not inside a git repository.
  }

  return simpleGit.gitP(options)
}

// Returns the remote of the current repository.
export const gitRemote = async (git: simpleGit.SimpleGit): Promise<string> => {
  const remotes = await git.getRemotes(true)
  if (remotes.length === 0) {
    throw new Error('No git remotes available')
  }

  for (const remote of remotes) {
    // We're trying to pick the remote called with the default git name 'origin'.
    if (remote.name === 'origin') {
      return stripCredentials(remote.refs.push)
    }
  }

  // Falling back to picking the first remote in the list if 'origin' is not found.
  return stripCredentials(remotes[0].refs.push)
}

// StripCredentials removes credentials from a remote HTTP url.
export const stripCredentials = (remote: string) => {
  try {
    const url = new URL(remote)
    url.username = ''
    url.password = ''

    return url.toString()
  } catch {
    return remote
  }
}

// Returns the hash of the current repository.
const gitHash = async (git: simpleGit.SimpleGit): Promise<string> => git.revparse('HEAD')

// Returns the tracked files of the current repository.
export const gitTrackedFiles = async (git: simpleGit.SimpleGit): Promise<string[]> => {
  const files = await git.raw('ls-files')

  return files.split(/\r\n|\r|\n/).filter((s) => s !== '')
}

// Returns the current hash, remote URL and tracked files paths.
export const getCommitInfo = async (
  git: simpleGit.SimpleGit,
  stdout: Writable,
  repositoryURL?: string
): Promise<CommitInfo | undefined> => {
  // Invoke git commands to retrieve the remote, hash and tracked files.
  // We're using Promise.all instead of Promive.allSettled since we want to fail early if
  // any of the promises fails.
  let remote: string
  let hash: string
  let trackedFiles: string[]
  try {
    if (repositoryURL) {
      ;[hash, trackedFiles] = await Promise.all([gitHash(git), gitTrackedFiles(git)])
      remote = repositoryURL
    } else {
      ;[remote, hash, trackedFiles] = await Promise.all([gitRemote(git), gitHash(git), gitTrackedFiles(git)])
    }
  } catch (e) {
    stdout.write(renderGitError(e))

    return
  }

  return new CommitInfo(hash, remote, trackedFiles)
}
