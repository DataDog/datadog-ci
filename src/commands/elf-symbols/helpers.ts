import {getBaseSourcemapIntakeUrl} from '../../helpers/base-intake-url'
import {RequestBuilder} from '../../helpers/interfaces'
import {MultipartPayload, upload, UploadOptions} from '../../helpers/upload'
import {getRequestBuilder} from '../../helpers/utils'

export const getElfRequestBuilder = (apiKey: string, cliVersion: string, site: string) =>
  getRequestBuilder({
    apiKey,
    baseUrl: getBaseSourcemapIntakeUrl(site),
    headers: new Map([
      ['DD-EVP-ORIGIN', 'datadog-ci_elf-symbols'],
      ['DD-EVP-ORIGIN-VERSION', cliVersion],
    ]),
    overrideUrl: 'api/v2/srcmap',
  })

// This function exists partially just to make mocking network calls easier.
export const uploadMultipartHelper = async (
  requestBuilder: RequestBuilder,
  payload: MultipartPayload,
  opts: UploadOptions
) => upload(requestBuilder)(payload, opts)
