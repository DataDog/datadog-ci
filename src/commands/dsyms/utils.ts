import {exec} from 'child_process'
import {promises} from 'fs'
import {tmpdir} from 'os'
import path from 'path'
import {promisify} from 'util'

import rimraf from 'rimraf'

import {buildPath} from '../../helpers/utils'

export const isZipFile = async (filepath: string) => {
  try {
    const stats = await promises.stat(filepath)

    return stats.size !== 0 && path.extname(filepath) === '.zip'
  } catch (error) {
    // Log to console "file exists yet empty" ?
    return false
  }
}

export const createUniqueTmpDirectory = async (): Promise<string> => {
  const uniqueValue = Math.random() * Number.MAX_SAFE_INTEGER
  const directoryPath = buildPath(tmpdir(), uniqueValue.toString())
  await promises.mkdir(directoryPath, {recursive: true})

  return directoryPath
}

export const deleteDirectory = async (directoryPath: string): Promise<void> =>
  new Promise((resolve, reject) => rimraf(directoryPath, () => resolve()))

export const zipDirectoryToArchive = async (directoryPath: string, archivePath: string) => {
  const cwd = path.dirname(directoryPath)
  const directoryName = path.basename(directoryPath)
  await execute(`zip -r '${archivePath}' '${directoryName}'`, cwd)
}

export const unzipArchiveToDirectory = async (archivePath: string, directoryPath: string) => {
  await promises.mkdir(directoryPath, {recursive: true})
  await execute(`unzip -o '${archivePath}' -d '${directoryPath}'`)
}

export const executeDwarfdump = async (dSYMPath: string): Promise<{stderr: string; stdout: string}> =>
  execute(`dwarfdump --uuid '${dSYMPath}'`)

export const executeLipo = async (
  objectPath: string,
  arch: string,
  newObjectPath: string
): Promise<{stderr: string; stdout: string}> => execute(`lipo '${objectPath}' -thin ${arch} -output '${newObjectPath}'`)

const execProc = promisify(exec)
const execute = (cmd: string, cwd?: string): Promise<{stderr: string; stdout: string}> =>
  execProc(cmd, {
    cwd,
    maxBuffer: 5 * 1024 * 5000,
  })

export const getBaseIntakeUrl = (datadogSite?: string) => {
  if (process.env.DATADOG_DSYM_INTAKE_URL) {
    return process.env.DATADOG_DSYM_INTAKE_URL
  } else if (datadogSite) {
    return 'https://sourcemap-intake.' + datadogSite
  }

  return 'https://sourcemap-intake.datadoghq.com'
}

export const pluralize = (nb: number, singular: string, plural: string) => {
  if (nb >= 2) {
    return `${nb} ${plural}`
  }

  return `${nb} ${singular}`
}
