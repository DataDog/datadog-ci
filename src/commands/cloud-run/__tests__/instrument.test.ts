// XXX temporary workaround for @google-cloud/run ESM/CJS module issues
import type {IContainer, IEnvVar, IVolumeMount} from '../types'

import {
  API_KEY_ENV_VAR,
  DATADOG_SITE_EU1,
  ENVIRONMENT_ENV_VAR,
  DD_TAGS_ENV_VAR,
  HEALTH_PORT_ENV_VAR,
  DD_LOG_LEVEL_ENV_VAR,
  LOGS_INJECTION_ENV_VAR,
  LOGS_PATH_ENV_VAR,
  SERVICE_ENV_VAR,
  SITE_ENV_VAR,
  DD_TRACE_ENABLED_ENV_VAR,
  VERSION_ENV_VAR,
} from '../../../constants'
import {makeRunCLI} from '../../../helpers/__tests__/testing-tools'
import * as apikey from '../../../helpers/apikey'
import * as instrumentHelpers from '../../../helpers/git/instrument-helpers'

import {InstrumentCommand} from '../instrument'
import * as cloudRunPromptModule from '../prompt'
import * as utils from '../utils'

jest.mock('../../../helpers/apikey')
jest.mock('../utils', () => ({
  ...jest.requireActual('../utils'),
  checkAuthentication: jest.fn(),
}))

const mockServicesClient = {
  servicePath: jest.fn(),
  getService: jest.fn(),
  updateService: jest.fn(),
}

jest.mock('@google-cloud/run', () => ({
  ServicesClient: jest.fn(() => mockServicesClient),
}))

describe('InstrumentCommand', () => {
  const runCLI = makeRunCLI(InstrumentCommand, ['cloud-run', 'instrument'])

  beforeEach(() => {
    jest.clearAllMocks()
    const mockValidator = {
      validateApiKey: jest.fn().mockResolvedValue(true),
      verifyApiKey: jest.fn().mockResolvedValue(undefined),
    }
    ;(apikey.newApiKeyValidator as jest.Mock).mockReturnValue(mockValidator)
    ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(true)

    // Reset mock client
    mockServicesClient.servicePath.mockImplementation(
      (project, region, service) => `projects/${project}/locations/${region}/services/${service}`
    )
  })

  describe('validates required variables', () => {
    test('should fail if project is missing', async () => {
      const {code, context} = await runCLI(['--services', 'test-service', '--region', 'us-central1'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('missing project')
    })

    test('should fail if services are missing', async () => {
      const {code, context} = await runCLI(['--project', 'test-project', '--region', 'us-central1'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('missing service(s)')
    })

    test('should fail if region is missing', async () => {
      const {code, context} = await runCLI(['--project', 'test-project', '--services', 'test-service'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('missing region')
    })
  })

  describe('main instrument command flow', () => {
    test('should fail if GCP credentials are invalid', async () => {
      ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(false)
      const {code, context} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
      ])
      expect(code).toBe(1)
      expect(context.stderr.toString()).toContain('Unable to authenticate with GCP')
    })

    test('should fail if sidecar instrumentation fails', async () => {
      const mockInstrumentSidecar = jest.fn().mockRejectedValue(new Error('Failed to instrument sidecar'))
      jest.spyOn(InstrumentCommand.prototype as any, 'instrumentSidecar').mockImplementation(mockInstrumentSidecar)

      const {code} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
      ])
      expect(code).toBe(1)
    })

    test('should succeed with valid parameters', async () => {
      const mockInstrumentSidecar = jest.fn().mockResolvedValue(undefined)
      jest.spyOn(InstrumentCommand.prototype as any, 'instrumentSidecar').mockImplementation(mockInstrumentSidecar)

      const {code} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
      ])
      expect(code).toBe(0)
      expect(mockInstrumentSidecar).toHaveBeenCalledWith('test-project', ['test-service'], 'us-central1', undefined)
    })
  })

  describe('snapshot tests', () => {
    const mockService = {
      name: 'projects/test-project/locations/us-central1/services/test-service',
      template: {
        containers: [
          {
            name: 'main-app',
            image: 'gcr.io/test-project/test-app:latest',
            env: [{name: 'NODE_ENV', value: 'production'}],
            volumeMounts: [],
          },
        ],
        volumes: [],
        revision: 'test-service-v1',
      },
    }

    beforeEach(() => {
      process.env[API_KEY_ENV_VAR] = 'test-api-key'
      process.env[SERVICE_ENV_VAR] = 'test-service'

      mockServicesClient.getService.mockResolvedValue([mockService])

      const mockOperation = {
        promise: jest.fn().mockResolvedValue([]),
      }
      mockServicesClient.updateService.mockResolvedValue([mockOperation])

      jest.restoreAllMocks()

      const mockValidator = {
        validateApiKey: jest.fn().mockResolvedValue(true),
        verifyApiKey: jest.fn().mockResolvedValue(undefined),
      }
      ;(apikey.newApiKeyValidator as jest.Mock).mockReturnValue(mockValidator)
      ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(true)

      // Re-apply git status mock after restoreAllMocks
      const mockGitStatus = jest.spyOn(instrumentHelpers as any, 'getCurrentGitStatus')
      mockGitStatus.mockImplementation(async () => ({
        ahead: 0,
        hash: '1be168ff837f043bde17c0314341c84271047b31',
        remote: 'git.repository_url:git@github.com:datadog/test.git',
        isClean: true,
        files: [],
      }))
    })

    test('prints dry run data with basic flags', async () => {
      const {code, context} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
        '--dry-run',
        '--env',
        'staging',
        '--version',
        '1.0.0',
        '--extra-tags',
        'team:backend,service:api',
        '--no-upload-git-metadata',
      ])

      expect(code).toBe(0)
      expect(context.stdout.toString()).toMatchSnapshot()
    })

    test('interactive mode', async () => {
      // Mock the prompts to return values
      jest.spyOn(cloudRunPromptModule, 'requestGCPProject').mockResolvedValue('interactive-project')
      jest.spyOn(cloudRunPromptModule, 'requestGCPRegion').mockResolvedValue('us-west1')
      jest.spyOn(cloudRunPromptModule, 'requestServiceName').mockResolvedValue('interactive-service')
      jest.spyOn(cloudRunPromptModule, 'requestSite').mockResolvedValue('datadoghq.com')
      jest.spyOn(cloudRunPromptModule, 'requestConfirmation').mockResolvedValue(true)

      // Mock the service for interactive mode
      const interactiveService = {
        ...mockService,
        name: 'projects/interactive-project/locations/us-west1/services/interactive-service',
      }
      mockServicesClient.getService.mockResolvedValue([interactiveService])

      const {code, context} = await runCLI(['--interactive', '--no-upload-git-metadata'])

      expect(code).toBe(0)
      expect(context.stdout.toString()).toMatchSnapshot()
    })
  })

  describe('createInstrumentedServiceConfig', () => {
    let command: InstrumentCommand

    beforeEach(() => {
      command = new InstrumentCommand()
      ;(command as any).tracing = undefined
      ;(command as any).sidecarImage = 'gcr.io/datadoghq/serverless-init:latest'
      ;(command as any).sidecarName = 'datadog-sidecar'
      ;(command as any).sharedVolumeName = 'shared-volume'
      ;(command as any).sharedVolumePath = '/shared-volume'
      ;(command as any).logsPath = '/shared-volume/logs/*.log'
    })

    test('adds sidecar and shared volume when missing', () => {
      const service = {
        template: {
          containers: [{name: 'main', env: [], volumeMounts: []}],
          volumes: [],
        },
      }

      const result = command.createInstrumentedServiceConfig(service, 'my-dd-service')

      // should have original + sidecar
      expect(result.template?.containers).toHaveLength(2)
      expect(result.template?.containers?.map((c: IContainer) => c.name)).toEqual(['main', 'datadog-sidecar'])

      // main container should get the shared volume mount
      const main = result.template?.containers?.find((c: IContainer) => c.name === 'main')
      expect(main?.volumeMounts?.some((vm: IVolumeMount) => vm.mountPath === '/shared-volume')).toBe(true)

      // should add the shared-volume
      expect(result.template?.volumes).toHaveLength(1)
      expect(result.template?.volumes?.[0].name).toBe('shared-volume')
    })

    test('does not add duplicate sidecar or volume when app and sidecar already present', () => {
      const appContainer = {
        name: 'app',
        env: [{name: SERVICE_ENV_VAR, value: 'old-service'}],
        volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
      }

      const sidecarContainer = {
        name: 'datadog-sidecar',
        env: [],
        volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
      }

      const existingVolume = {name: 'shared-volume', emptyDir: {}}

      const service = {
        template: {
          containers: [appContainer, sidecarContainer],
          volumes: [existingVolume],
        },
      }

      const result = command.createInstrumentedServiceConfig(service, 'my-dd-service')

      // should not add another sidecar
      expect(result.template?.containers).toHaveLength(2)
      expect(result.template?.containers?.map((c: IContainer) => c.name)).toEqual(['app', 'datadog-sidecar'])

      // should not add another shared-volume
      expect(result.template?.volumes).toHaveLength(1)
      expect(result.template?.volumes?.[0].name).toBe('shared-volume')
    })

    test('uses custom configuration values', () => {
      ;(command as any).sidecarImage = 'custom-image:v1.0'
      ;(command as any).sidecarName = 'custom-sidecar'
      ;(command as any).sharedVolumeName = 'custom-volume'
      ;(command as any).sharedVolumePath = '/custom/path'
      ;(command as any).logsPath = '/custom/path/logs/*.log'
      ;(command as any).sidecarCpus = '2'
      ;(command as any).sidecarMemory = '256Mi'

      const service = {
        template: {
          containers: [{name: 'main', env: [], volumeMounts: []}],
          volumes: [],
        },
      }

      const result = command.createInstrumentedServiceConfig(service, 'test-service')

      // Check sidecar container has custom values
      const sidecarContainer = result.template?.containers?.find((c: IContainer) => c.name === 'custom-sidecar')
      expect(sidecarContainer).toBeDefined()
      expect(sidecarContainer?.image).toBe('custom-image:v1.0')
      expect(sidecarContainer?.name).toBe('custom-sidecar')
      expect(sidecarContainer?.volumeMounts?.[0]?.name).toBe('custom-volume')
      expect(sidecarContainer?.volumeMounts?.[0]?.mountPath).toBe('/custom/path')
      expect(sidecarContainer?.env?.find((e: IEnvVar) => e.name === 'DD_SERVERLESS_LOG_PATH')?.value).toBe(
        '/custom/path/logs/*.log'
      )
      expect(sidecarContainer?.resources?.limits?.cpu).toBe('2')
      expect(sidecarContainer?.resources?.limits?.memory).toBe('256Mi')

      // Check main container has custom volume mount
      const mainContainer = result.template?.containers?.find((c: IContainer) => c.name === 'main')
      expect(mainContainer?.volumeMounts?.[0]?.name).toBe('custom-volume')
      expect(mainContainer?.volumeMounts?.[0]?.mountPath).toBe('/custom/path')

      // Check custom volume is created
      expect(result.template?.volumes?.[0]?.name).toBe('custom-volume')
    })
  })

  describe('buildSidecarContainer', () => {
    let command: InstrumentCommand

    beforeEach(() => {
      command = new InstrumentCommand()
      ;(command as any).tracing = undefined
      ;(command as any).sidecarImage = 'gcr.io/datadoghq/serverless-init:latest'
      ;(command as any).sidecarName = 'datadog-sidecar'
      ;(command as any).sharedVolumeName = 'shared-volume'
      ;(command as any).sharedVolumePath = '/shared-volume'
      ;(command as any).logsPath = '/shared-volume/logs/*.log'
    })

    test('custom flags set correct env vars', () => {
      ;(command as any).environment = 'dev'
      ;(command as any).version = 'v123.456'
      ;(command as any).logLevel = 'debug'
      ;(command as any).llmobs = 'my-llm-app'
      ;(command as any).extraTags = 'foo:bar,abc:def'

      const newSidecarContainer = command.buildSidecarContainer(undefined, 'my-service')
      const expected: IEnvVar[] = [
        {name: SERVICE_ENV_VAR, value: 'my-service'},
        {name: ENVIRONMENT_ENV_VAR, value: 'dev'},
        {name: VERSION_ENV_VAR, value: 'v123.456'},
        {name: SITE_ENV_VAR, value: 'datadoghq.com'},
        {name: LOGS_PATH_ENV_VAR, value: (command as any).logsPath as string},
        {name: API_KEY_ENV_VAR, value: process.env.DD_API_KEY ?? ''},
        {name: HEALTH_PORT_ENV_VAR, value: '5555'},
        {name: LOGS_INJECTION_ENV_VAR, value: 'true'},
        {name: DD_TRACE_ENABLED_ENV_VAR, value: 'true'},
        {name: DD_LOG_LEVEL_ENV_VAR, value: 'debug'},
        {name: DD_TAGS_ENV_VAR, value: 'foo:bar,abc:def'},
      ]
      expect(newSidecarContainer.env).toEqual(expect.arrayContaining(expected))
      expect(newSidecarContainer.env).toHaveLength(expected.length)
    })

    test('overwrites intended env vars; leaves existing env vars unchanged', () => {
      process.env[API_KEY_ENV_VAR] = 'mock-api-key'
      const existingSidecarContainer = {
        name: 'datadog-sidecar',
        env: [
          // Following env vars should be left unchanged
          {name: SITE_ENV_VAR, value: DATADOG_SITE_EU1},
          {name: LOGS_PATH_ENV_VAR, value: 'some-log-path'},
          {name: LOGS_INJECTION_ENV_VAR, value: 'false'},
          {name: DD_TRACE_ENABLED_ENV_VAR, value: 'false'},
          {name: HEALTH_PORT_ENV_VAR, value: '12345'},
          {name: 'CUSTOM_ENV_VAR', value: 'some-value'},
          // Following env vars should be overwritten
          {name: API_KEY_ENV_VAR, value: '123'},
          {name: SERVICE_ENV_VAR, value: 'old-service'},
        ],
        volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
      }
      const newSidecarContainer = command.buildSidecarContainer(existingSidecarContainer, 'new-service')
      const expected: IEnvVar[] = [
        {name: SITE_ENV_VAR, value: DATADOG_SITE_EU1},
        {name: LOGS_PATH_ENV_VAR, value: '/shared-volume/logs/*.log'},
        {name: LOGS_INJECTION_ENV_VAR, value: 'false'},
        {name: DD_TRACE_ENABLED_ENV_VAR, value: 'false'},
        {name: HEALTH_PORT_ENV_VAR, value: '12345'},
        {name: 'CUSTOM_ENV_VAR', value: 'some-value'},
        {name: API_KEY_ENV_VAR, value: 'mock-api-key'},
        {name: SERVICE_ENV_VAR, value: 'new-service'},
      ]
      for (const expectedEnv of expected) {
        const actual = newSidecarContainer.env?.find((value) => value.name === expectedEnv.name)
        expect(actual?.value).toBe(expectedEnv.value)
      }
    })
  })
})
