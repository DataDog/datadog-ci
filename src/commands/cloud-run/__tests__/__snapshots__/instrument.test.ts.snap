// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstrumentCommand snapshot tests interactive mode 1`] = `
"
üê∂ Instrumenting Cloud Run service(s)

[!] No DD_SERVICE env var found. Will default to the service name.
‚úî Required flags verified

üîë Verifying GCP credentials...
‚úî GCP credentials verified!


‚¨áÔ∏è Fetching existing service configurations from Cloud Run...

üöÄ Instrumenting Cloud Run services with sidecar...
  {
    "name": "projects/interactive-project/locations/us-west1/services/interactive-service",
    "template": {
      "containers": [
        {
          "env": [
            {
              "name": "NODE_ENV",
              "value": "production"
+           },
+           {
+             "name": "DD_LOGS_INJECTION",
+             "value": "true"
+           },
+           {
+             "name": "DD_SERVICE",
+             "value": "interactive-service"
+           },
+           {
+             "name": "DD_API_KEY",
+             "value": "PLACEHOLDER"
            }
          ],
          "image": "gcr.io/test-project/test-app:latest",
          "name": "main-app",
-         "volumeMounts": []
+         "volumeMounts": [
+           {
+             "mountPath": "/shared-volume",
+             "name": "shared-volume"
+           }
+         ]
+       },
+       {
+         "env": [
+           {
+             "name": "DD_SITE",
+             "value": "datadoghq.com"
+           },
+           {
+             "name": "DD_LOGS_INJECTION",
+             "value": "true"
+           },
+           {
+             "name": "DD_TRACE_ENABLED",
+             "value": "true"
+           },
+           {
+             "name": "DD_HEALTH_PORT",
+             "value": "5555"
+           },
+           {
+             "name": "DD_API_KEY",
+             "value": "PLACEHOLDER"
+           },
+           {
+             "name": "DD_SERVICE",
+             "value": "interactive-service"
+           },
+           {
+             "name": "DD_TAGS",
+             "value": "git.commit.sha:1be168ff837f043bde17c0314341c84271047b31,git.repository_url:git.repository_url:github.com/datadog/test.git"
+           },
+           {
+             "name": "DD_SERVERLESS_LOG_PATH",
+             "value": "/shared-volume/logs/*.log"
+           }
+         ],
+         "image": "gcr.io/datadoghq/serverless-init:latest",
+         "name": "datadog-sidecar",
+         "resources": {
+           "limits": {
+             "cpu": "1",
+             "memory": "512Mi"
+           }
+         },
+         "startupProbe": {
+           "failureThreshold": 3,
+           "initialDelaySeconds": 0,
+           "periodSeconds": 10,
+           "tcpSocket": {
+             "port": "5555"
+           },
+           "timeoutSeconds": 1
+         },
+         "volumeMounts": [
+           {
+             "mountPath": "/shared-volume",
+             "name": "shared-volume"
+           }
+         ]
        }
      ],
-     "revision": "test-service-v1",
-     "volumes": []
+     "volumes": [
+       {
+         "emptyDir": {
+           "medium": 1
+         },
+         "name": "shared-volume"
+       }
+     ]
    }
  }
‚úÖ Cloud Run instrumentation completed successfully!
"
`;

exports[`InstrumentCommand snapshot tests prints dry run data with basic flags 1`] = `
"
[Dry Run] üê∂ Instrumenting Cloud Run service(s)

[!] No DD_SERVICE env var found. Will default to the service name.
‚úî Required flags verified

üîë Verifying GCP credentials...
‚úî GCP credentials verified!


[Dry Run] ‚¨áÔ∏è Fetching existing service configurations from Cloud Run...

[Dry Run] üöÄ Instrumenting Cloud Run services with sidecar...
  {
    "name": "projects/test-project/locations/us-central1/services/test-service",
    "template": {
      "containers": [
        {
          "env": [
            {
              "name": "NODE_ENV",
              "value": "production"
+           },
+           {
+             "name": "DD_LOGS_INJECTION",
+             "value": "true"
+           },
+           {
+             "name": "DD_SERVICE",
+             "value": "test-service"
+           },
+           {
+             "name": "DD_API_KEY",
+             "value": "PLACEHOLDER"
            }
          ],
          "image": "gcr.io/test-project/test-app:latest",
          "name": "main-app",
-         "volumeMounts": []
+         "volumeMounts": [
+           {
+             "mountPath": "/shared-volume",
+             "name": "shared-volume"
+           }
+         ]
+       },
+       {
+         "env": [
+           {
+             "name": "DD_SITE",
+             "value": "datadoghq.com"
+           },
+           {
+             "name": "DD_LOGS_INJECTION",
+             "value": "true"
+           },
+           {
+             "name": "DD_TRACE_ENABLED",
+             "value": "true"
+           },
+           {
+             "name": "DD_HEALTH_PORT",
+             "value": "5555"
+           },
+           {
+             "name": "DD_API_KEY",
+             "value": "PLACEHOLDER"
+           },
+           {
+             "name": "DD_SERVICE",
+             "value": "test-service"
+           },
+           {
+             "name": "DD_ENV",
+             "value": "staging"
+           },
+           {
+             "name": "DD_VERSION",
+             "value": "1.0.0"
+           },
+           {
+             "name": "DD_TAGS",
+             "value": "team:backend,service:api,git.commit.sha:1be168ff837f043bde17c0314341c84271047b31,git.repository_url:git.repository_url:github.com/datadog/test.git"
+           },
+           {
+             "name": "DD_SERVERLESS_LOG_PATH",
+             "value": "/shared-volume/logs/*.log"
+           }
+         ],
+         "image": "gcr.io/datadoghq/serverless-init:latest",
+         "name": "datadog-sidecar",
+         "resources": {
+           "limits": {
+             "cpu": "1",
+             "memory": "512Mi"
+           }
+         },
+         "startupProbe": {
+           "failureThreshold": 3,
+           "initialDelaySeconds": 0,
+           "periodSeconds": 10,
+           "tcpSocket": {
+             "port": "5555"
+           },
+           "timeoutSeconds": 1
+         },
+         "volumeMounts": [
+           {
+             "mountPath": "/shared-volume",
+             "name": "shared-volume"
+           }
+         ]
        }
      ],
-     "revision": "test-service-v1",
-     "volumes": []
+     "volumes": [
+       {
+         "emptyDir": {
+           "medium": 1
+         },
+         "name": "shared-volume"
+       }
+     ]
    }
  }

[Dry Run] Would have updated service test-service with the above changes.
"
`;
